### PR TITLE
Detect time comparisons and rename linter to badtime

### PR DESCRIPTION
**Pull request by richardartoul**
This changes makes the linter support detecting stuff like time.Time == time.Time

I feel like the linter should probably be renamed badtime, but I don't know if I want to deal with updating ci-scripts/m3x/m3db repos right now...

### CODE REVIEW

File: linters/badtime/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Badtime
```
Comment (richardartoul) [inline linters/badtime/README.md]: I'm not sure why its showing so many changes to the readme, I just replaced maptime with badtime and added a few more sentences explaining the expanded functionality

Comment (richardartoul) [reply linters/badtime/README.md]: Oh its because I renamed the file, thats why

File: README.md [hunk 2]
```diff
@@ -3,4 +3,4 @@ Various build tools used as part of the M3DB project
 
 ## Contents
 
-1. [maptime](https://github.com/m3db/build-tools/blob/master/linters/maptime/README.md) - gometalinter plugin that detects Golang maps which have an instance of time.Time as part of the key
\ No newline at end of file
+1. [badtime](https://github.com/m3db/build-tools/blob/master/linters/badtime/README.md) - gometalinter plugin that detects inappropriate usage of the time.Time struct
```
Comment (prateek) [inline README.md]: Love the new name :)

File: linters/badtime/main.go [hunk 3]
```diff
@@ -28,19 +28,21 @@ import (
...
-	"path"
 	"strings"
 
 	"golang.org/x/tools/go/loader"
 
 	"github.com/kisielk/gotool"
 )
 
-type callbackFunc func(loc token.Position, keyStr, valStr string)
+type mapCallbackFunc func(loc token.Position, keyStr, valStr string)
```
Comment (prateek) [inline linters/badtime/main.go]: looks like these have the same signature, would it be a better abstraction to share the type? 

something like:
```
type validationCallbackFn func(loc token.Position, keyStr, valStr string)
```

File: linters/badtime/main.go [hunk 4]
```diff
@@ -98,46 +107,66 @@ func handleImportPaths(importPaths []string, buildTags []string, callback callba
...
-	}
-}
-
 type mapVisitor struct {
-	fs       *token.FileSet
-	types    map[ast.Expr]types.TypeAndValue
-	callback callbackFunc
+	fs               *token.FileSet
+	types            map[ast.Expr]types.TypeAndValue
+	mapCallback      mapCallbackFunc
```
Comment (prateek) [inline linters/badtime/main.go]: maybe change this to take a list of callbacks, might be more compose-able that way?

Comment (prateek) [reply linters/badtime/main.go]: actually disregard

File: linters/badtime/main.go (hunk 3 above)
Comment (prateek) [reply linters/badtime/main.go]: see why you've kept them separate, disregard.

File: linters/badtime/main.go [hunk 5]
```diff
@@ -98,46 +107,66 @@ func handleImportPaths(importPaths []string, buildTags []string, callback callba
...
+	equalityCallback equalityCallbackFunc
 }
 
 func (v mapVisitor) Visit(node ast.Node) ast.Visitor {
+	// Detect time.Time == time.Time
```
Comment (prateek) [inline linters/badtime/main.go]: do you want to split this into two separate visitors? weird that you're using `mapVisitor` to check binary expressions.

Comment (richardartoul) [reply linters/badtime/main.go]: I think its a little cleaner (and less wasteful) as one visitor. I'll just rename it

**Pull request merged**